Python/chapter03/3.6 - Animal Shelter/Nick.py

```diff
@@ -0,0 +1,88 @@
+#!/usr/bin/python3
+'''Animal Shelter: An animal shelter, which holds only
+dogs and cats, operates on a strictly"first in, first
+out" basis. People must adopt either the "oldest"
+(based on arrival time) of all animals at the shelter,
+or they can select whether they would prefer a dog or
+a cat (and will receive the oldest animal of that type).
+They cannot select which specific animal they would
+like. Create the data structures to maintain this
+system and implement operations such as enqueue,
+dequeueAny, dequeueDog, and dequeueCat. You may use
+the built-in Linked list data structure.
+Hints: #22, #56, #63'''
+import unittest
+
+
+class ShelterQueue:
+    def __init__(self):
+        self.queue = []
+
+    def enqueue(self, animal):
+        if len(self.queue) == 0:
+            self.queue.append(animal)
+        else:
+            self.queue.insert(0, animal)
+
+    def printQueue(self):
+        print('----- Beginning of Queue ------')
+        for x in self.queue[::-1]:
+            print(x)
+        print('------ End of Queue ------')
+
+    def dequeueAnimal(self, animal):
+        if len(self.queue) == 0:
+            return
+        temp = False
+        for x in self.queue[::-1]:
```

Comment:
here you can use the [guard clause pattern](https://refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html) instead of moving the condition to a different block:

```python3
for pet in self.queue[::-1]:
  if not pet.startswith(animal_type):
    continue
  self.queue.remove(pet)
  return pet
return None
```